src/mjolnir/ferry_connections.cc

```diff
@@ -131,13 +131,21 @@ std::pair<uint32_t, bool> ShortestPath(const uint32_t start_node_idx,
       // Have seen cases where the immediate connections are high class roads
       // but then there are service roads (lanes) immediately after (like
       // Twawwassen Terminal near Vancouver,BC)
-      if (GetBestNonFerryClass(expanded_bundle.node_edges) <= kFerryUpClass) {
+      //
+      // Improved termination: Continue expanding to find more stable paths
+      // and handle sequences like ferry -> highway -> service -> highway
+      uint32_t current_best_class = GetBestNonFerryClass(expanded_bundle.node_edges);
+      if (current_best_class <= kFerryUpClass) {
         // Set the last label index - shortest path is recovered backwards from this
         // label to the ferry start
         last_label_idx = label_idx;
 
-        // TODO - better termination criteria?!
-        if (n > 400) {
+        // Improved termination criteria: Don't stop immediately after finding
```

Comment:
I guess "improved" compared to what it was before. Rewrite comment to represent only what code below does